pennylane/ops/qubit/matrix_ops.py

```diff
@@ -296,9 +319,92 @@ def compute_decomposition(U: TensorLike, wires: WiresLike):
                 raise DecompositionUndefinedError(
                     "The decomposition of a two-qubit sparse QubitUnitary is undefined."
                 )
+
             return qml.ops.two_qubit_decomposition(U, Wires(wires))
 
-        return super(QubitUnitary, QubitUnitary).compute_decomposition(U, wires=wires)
+        ops = []
+
+        # pylint: disable=import-outside-toplevel
+        try:
+            # Wrap scipy's cossin function with pure_callback to make the jit decomposition compatible
+
+            import jax
+
+            def scipy_cossin_callback(U_flat, p, q):
+                dim = int(np.sqrt(U_flat.size))
+                U_np = U_flat.reshape((dim, dim))
+                (u1, u2), theta, (v1_dagg, v2_dagg) = sp.linalg.cossin(
+                    U_np, p=p, q=q, separate=True
+                )
+                return u1, u2, theta, v1_dagg, v2_dagg
+
+            def cossin_decomposition(U, p, q):
+                dtype = U.dtype
+                U_flat = U.reshape(-1)
+
+                def callback(U_flat):
+                    return tuple(
+                        arr.astype(dtype) for arr in scipy_cossin_callback(np.asarray(U_flat), p, q)
+                    )
+
+                u1, u2, theta, v1_dagg, v2_dagg = jax.pure_callback(
+                    callback,
+                    result_shape_dtypes=(
+                        jax.ShapeDtypeStruct((p, p), dtype),
+                        jax.ShapeDtypeStruct((p, p), dtype),
+                        jax.ShapeDtypeStruct((p,), dtype),
+                        jax.ShapeDtypeStruct((q, q), dtype),
+                        jax.ShapeDtypeStruct((q, q), dtype),
+                    ),
+                    U_flat=U_flat,
+                )
+
+                return (u1, u2), theta, (v1_dagg, v2_dagg)
+
+        except ImportError:
+
+            def cossin_decomposition(U, p, q):
+                return sp.linalg.cossin(U, p=p, q=q, separate=True)
+
+        # Combining the two equalities in Fig. 14 [https://arxiv.org/pdf/quant-ph/0504100], we can express
+        # a n-qubit unitary U with four (n-1)-qubit unitaries and three multiplexed rotations ( via `qml.SelectPauliRot`)
+
+        p = 2 ** (len(wires) - 1)
+        q = 2 ** (len(wires) - 1)
+
+        (u1, u2), theta, (v1_dagg, v2_dagg) = cossin_decomposition(U, p, q)
```

Comment:
Do we need two variables here?
```suggestion
        p = 2 ** (len(wires) - 1)

        (u1, u2), theta, (v1_dagg, v2_dagg) = cossin_decomposition(U, p, p)
```

Reply:
If that's the case, does `cossin_decomposition` need two length variables? Might just trim `q` from there as well.

Reply:
removed "q" from the code 👍